src/EthernetClient.cpp

```diff
@@ -96,17 +96,34 @@ size_t EthernetClient::write(const uint8_t *buf, size_t size) {
     return 0;
   }
 
-  if(ERR_OK != tcp_write(_tcp_client->pcb, buf, size, TCP_WRITE_FLAG_COPY)) {
-    return 0;
-  }
+  size_t max_send_size, bytes_to_send;
+  size_t bytes_sent = 0;
+  size_t bytes_left = size;
+  err_t res;
+
+  while(bytes_sent != size){
+    max_send_size = tcp_sndbuf(_tcp_client->pcb);
+    bytes_to_send = bytes_left > max_send_size ? max_send_size : bytes_left;
+
+    if(bytes_to_send > 0){
```

Comment:
Instead of a timeout implementation, you can return 0 when bytes_to_send  == 0;
So if there is no buffer free it is up to application to manage (retransmition, ...).
for example:
add an else to 
```
if(bytes_to_send > 0) {
...
 } 
else {
  return 0;
}
```
and thus we avoid inifnite loop

Reply:
I cannot find exactly what behavior the write function should have but I interpreted it as it should block?

* The [official documentation](https://www.arduino.cc/en/Reference/ClientWrite) says it's  optional to read the number of written bytes and the only way you could ignore it would be if you knew everything was sent.
* [This MQTT](https://github.com/256dpi/arduino-mqtt/blob/master/src/MQTTClient.cpp#L75) implementation requires it to block until all is sent and it would not work if only a chunk was sent with the write call. The MQTT library could of course be broken.
* the [Arduino Ethernet shield](https://github.com/arduino-libraries/Ethernet/blob/master/src/socket.cpp#L422) documentation says 0 on failure, 1 on success but truncates the buffer if it's too large and returns the number of bytes that were sent.

It feels like this part of the Arduino isn't well documentated/specified? What to do? 

The proper solution would be to copy the behavior of [Linux' write](http://man7.org/linux/man-pages/man2/write.2.html) but that would require changes of the main Arduino code/documentation itself.

Reply:
Hi @ramboerik,
Ok, that is fine for me if you keep your original proposal.
What about other remarks ?

Reply:
Ok, I keep the blocking behavior if no one else has any opinion.